File: DesignPattern/creational_pattern/builder_pattern.cpp
#include <iostream>
#include <string>

// Product
class Computer {
private:
    std::string cpu_;
    int ram_ = 0;
    std::string storage_;

public:
    void setCPU(const std::string& cpu) {
        cpu_ = cpu;
    }

    void setRAM(int ram) {
        ram_ = ram;
    }

    void setStorage(const std::string& storage) {
        storage_ = storage;
    }

    void showSpecs() {
        std::cout << "CPU: " << cpu_ << ", RAM: " << ram_ << "GB, Storage: " << storage_ << std::endl;
    }
};

// Abstract Builder
class ComputerBuilder {
public:
    virtual void buildCPU() = 0;
    virtual void buildRAM() = 0;
    virtual void buildStorage() = 0;
    virtual Computer getResult() = 0;
};

// Concrete Builder for Gaming Computer
class GamingComputerBuilder : public ComputerBuilder {
public:
    void buildCPU() override {
        computer_.setCPU("Intel Core i9");
    }

    void buildRAM() override {
        computer_.setRAM(32);
    }

    void buildStorage() override {
        computer_.setStorage("2TB SSD");
    }

    Computer getResult() override {
        return computer_;
    }

private:
    Computer computer_;
};

// Concrete Builder for Office Computer
class OfficeComputerBuilder : public ComputerBuilder {
public:
    void buildCPU() override {
        computer_.setCPU("Intel Core i7");
    }

    void buildRAM() override {
        computer_.setRAM(16);
    }

    void buildStorage() override {
        computer_.setStorage("1TB HDD");
    }

    Computer getResult() override {
        return computer_;
    }

private:
    Computer computer_;
};

// Director
class ComputerDirector {
public:
    void setBuilder(ComputerBuilder* builder) {
        builder_ = builder;
    }

    Computer construct() {
        builder_->buildCPU();
        builder_->buildRAM();
        builder_->buildStorage();
        return builder_->getResult();
    }

private:
    ComputerBuilder* builder_;
};

int main() {
    GamingComputerBuilder gamingBuilder;
    OfficeComputerBuilder officeBuilder;
    ComputerDirector director;

    director.setBuilder(&gamingBuilder);
    Computer gamingComputer = director.construct();
    std::cout << "Gaming Computer Specs:" << std::endl;
    gamingComputer.showSpecs();

    director.setBuilder(&officeBuilder);
    Computer officeComputer = director.construct();
    std::cout << "Office Computer Specs:" << std::endl;
    officeComputer.showSpecs();

    return 0;
}




/*
- The Builder Pattern is a creational design pattern that allows you to construct complex objects step by step. It separates the construction 
of a complex object from its representation, allowing you to create different representations of an object by using the same construction process. 
This pattern is useful when you need to create objects with many optional components or configurations.

- In this example: Computer is the product class that you want to build.

- ComputerBuilder is an abstract builder class with methods for building various components of the Computer.

- GamingComputerBuilder and OfficeComputerBuilder are concrete builder classes that inherit from ComputerBuilder. They implement 
the methods to build components with specific configurations for gaming and office computers.

- ComputerDirector is responsible for directing the construction process using a specific builder.

- In the main function, we create instances of the GamingComputerBuilder and OfficeComputerBuilder, and then use the ComputerDirector 
to construct gaming and office computers with different configurations. The builder pattern allows you to create complex objects while keeping 
the construction process and product separate, making it easy to create different variants of a product.
*/

File: DesignPattern/creational_pattern/builder_pattern.py
class Computer:
    def __init__(self, cpu, ram, storage):
        self.cpu = cpu
        self.ram = ram
        self.storage = storage

    def __str__(self):
        return f"CPU: {self.cpu}, RAM: {self.ram}GB, Storage: {self.storage}"

class ComputerBuilder:
    def __init__(self):
        self.cpu = None
        self.ram = None
        self.storage = None

    def set_cpu(self, cpu):
        self.cpu = cpu
        return self

    def set_ram(self, ram):
        self.ram = ram
        return self

    def set_storage(self, storage):
        self.storage = storage
        return self

    def build(self):
        return Computer(self.cpu, self.ram, self.storage)

# Director
class ComputerDirector:
    def construct_gaming_computer(self):
        return ComputerBuilder() \
            .set_cpu("Intel Core i9") \
            .set_ram(32) \
            .set_storage("2TB SSD") \
            .build()

    def construct_office_computer(self):
        return ComputerBuilder() \
            .set_cpu("Intel Core i7") \
            .set_ram(16) \
            .set_storage("1TB HDD") \
            .build()

if __name__ == "__main__":
    director = ComputerDirector()

    gaming_computer = director.construct_gaming_computer()
    print("Gaming Computer Specs:")
    print(gaming_computer)

    office_computer = director.construct_office_computer()
    print("Office Computer Specs:")
    print(office_computer)



'''
Output:

Gaming Computer Specs:
CPU: Intel Core i9, RAM: 32GB, Storage: 2TB SSD
Office Computer Specs:
CPU: Intel Core i7, RAM: 16GB, Storage: 1TB HDD
'''

'''
- The Builder Pattern in Python is a creational design pattern that allows you to construct complex objects step by step. It separates 
the construction of a complex object from its representation, enabling the creation of different representations of an object by using 
the same construction process. This pattern is useful when you need to create objects with many optional components or configurations.

- In this example: Computer is the product class that you want to build.
- ComputerBuilder is the builder class with methods for setting different components of the computer. Each method returns self, allowing for method chaining.
- The ComputerDirector is responsible for directing the construction process using specific builder methods to build gaming and office computers.

- In the main section, we create an instance of the ComputerDirector and use it to construct gaming and office computers with different configurations. 
The builder pattern allows you to create complex objects while keeping the construction process and product separate, making it easy to create different variants of a product.
'''